### ISSUE TITLE
梯度累积中的问题

### ISSUE THREAD

**Issue by ChaoLi977** [OFA-Sys/Chinese-CLIP#162]
关于梯度累积的代码，有几点不太明白，想请教一下。
1. accum_image_features 和accum_text_features已经得到了，为什么在get_loss中又要重新计算一个batch的特征。这样的话，所有的特征都被计算了两次，造成资源的浪费
2.  get_loss进行了accum_freq次，但是每次的accum_image_features和accum_text_features并没有发生变化。loss的计算是否是重复了？

相关代码如下：

            # First, cache the features without any gradient tracking.
            with torch.no_grad():
                with autocast(enabled=(args.precision == "amp")):
                    chunk_image_features, chunk_text_features, _ = model(images, texts)
                accum_image_features.append(chunk_image_features)
                accum_text_features.append(chunk_text_features)

                accum_images.append(images)
                accum_texts.append(texts)

            # If (i + 1) % accum_freq is not zero, move on to the next batch.
            if ((i + 1) % args.accum_freq) > 0:
                # FIXME this makes data time logging unreliable when accumulating
                continue

            # Now, ready to take gradients for the last accum_freq batches.
            # Re-do the forward pass for those batches, and use the cached features from the other batches as negatives.
            # Call backwards each time, but only step optimizer at the end.
            optimizer.zero_grad()
            for j in range(args.accum_freq):
                images = accum_images[j]
                texts = accum_texts[j]
                with autocast(enabled=(args.precision == "amp")):
                    # `total_loss` and `acc` are coarsely sampled, taking only the last result in the loop.
                    # Although each result should be the same in theory, it will be slightly different in practice
                    total_loss, acc = get_loss(model, images, texts, loss_img, loss_txt, args, accum_image_features, accum_text_features, j)
                if args.precision == "amp":
                    scaler.scale(total_loss).backward()
                else:
                    total_loss.backward()

            if args.precision == "amp":
                scaler.step(optimizer)
                scaler.update()
            else:
                optimizer.step()
`

**Comment by DtYXs** [comment]
您好，由于梯度累积通常是在GPU显存有限的情形下应用，而`accum_image_features `和`accum_text_features`是不带有梯度的（在`with torch.no_grad()`）中，所以`get_loss`进行`accum_freq`次分别计算每一部分的梯度，`accum_image_features`和`accum_text_features`让每个batch的数据见到更多样本来模拟更大的batchsize。具体的细节可以参考`get_loss`函数内部~
